Reprise: new lint: Unintentional return of unit from closures expecting Ord

Uriopass opened the pull request:
This lint catches cases where the last statement of a closure expecting
an instance of Ord has a trailing semi-colon. It compiles since the
closure ends up return () which also implements Ord but causes
unexpected results in cases such as sort_by_key.

Fixes #5080

Reprise of #5348 where I addressed all the comments there

changelog: add lint [`unit_return_expecting_ord`]

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Manishearth (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md) for more information.

Uriopass commented:
r? @flip1995 since he worked on the previous pr

Uriopass commented:
(Looks like the CI doesn't work but it seems unrelated to the changes ?)

flip1995 commented:
Yes this is unrelated and will be fixed tomorrow (if no other unexpected breakage occurs). In regard of reviewing this: I will do that after we get Clippy working again in the Clippy repo and in the rustc repo. (hopefully in one or two days)

flip1995 commented:
What you can do in the meantime: Please squash the commit b8d55fc in your rebase commit, since the rebase commits reverts the former.

Uriopass commented:
Squashed

Uriopass commented:
Looks like there are unrelated errors again, and a related one which I'll fix after the other errors are fixed

Uriopass commented:
rebased

Uriopass commented:
Ok now I have actual test fails however I have no idea where it comes from :(

flip1995 commented on clippy_lints/src/unintentional_unit_return.rs:
```diff
@@ -0,0 +1,158 @@
+fn get_args_to_check<'tcx>(cx: &LateContext<'_, 'tcx>, expr: &'tcx Expr<'tcx>) -> Vec<(usize, String)> {
+    let mut args_to_check = Vec::new();
+    if let Some(def_id) = cx.tables.type_dependent_def_id(expr.hir_id) {
+        let fn_sig = cx.tcx.fn_sig(def_id);
```
This produces an ICE. Probably because the `def_id` doesn't belong to a function signature?

Uriopass replied:
Ah ! Indeed it's called for every expression

flip1995 commented on clippy_lints/src/unintentional_unit_return.rs:
```diff
@@ -0,0 +1,158 @@
+                        if let Some(return_ty_pred) = get_projection_pred(cx, generics, trait_pred);
+                        then {
+                            if ord_preds.iter().any(|ord|  {
+                                unwrap_trait_pred(cx, ord).unwrap().self_ty()  == return_ty_pred.ty
```
This looks fishy. if this is safe to unwrap, please use expect, with a explanation instead.

Uriopass replied:
That was indeed a mistake, simplifying the code removes the unwrap

Uriopass commented:
(I'm sorry for wasting the CI a bit, I never contributed to clippy and can't run dogfood test since I'm on windows)

Uriopass commented:
:tada:

Uriopass commented:
Also addressing original author's remark:

`If returning unit is intentional, then there is no way of specifying this without triggering needless_return lint`

This is true for lots of lints, but they can be disabled selectively using `#[allow(clippy::unintentional_unit_return)]` no ?
I'm also not sure when lints should be in the nursery or not. Should this one be ?

flip1995 commented on tests/ui/unintentional_unit_return.rs:
```diff
@@ -0,0 +1,21 @@
+    structs.sort_by_key(|s| double(s.field));
+    structs.is_sorted_by_key(|s| {
+        double(s.field);
+    });
```
Can you add a check for a return statement in the closure:
```rust
structs.sort_by_key(|key| {
    return double(s.field);
});
```

This should not lint, since it works as expected: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d4a33a63a3feef6dbbd9e10bafc7f777)

Uriopass replied:
It does work, but it lints on "needless return"

flip1995 replied:
Please add a test for this, but add an ``allow`` for the needless return lint.

Uriopass replied:
Yes that's what I did :-) Waiting for rustup

flip1995 commented on clippy_lints/src/unintentional_unit_return.rs:
```diff
@@ -0,0 +1,145 @@
+        if let ExprKind::Closure(_, _fn_decl, body_id, _span, _) = arg.kind;
+        let body = cx.tcx.hir().body(body_id);
+        if let ExprKind::Block(block, _) = body.value.kind;
+        if block.expr.is_none();
+        if let Some(stmt) = block.stmts.last();
+        if let StmtKind::Semi(_) = stmt.kind;
```
I think the safer way would be to check the inferred return type of the closure. You can try if you can use `utils::return_ty` to get the return type of the closure. If it is `()`, then lint.

Uriopass replied:
return_ty doesn't work for closures :-(

Uriopass replied:
However I do not know how to get the substitution

Uriopass replied:
Done! Although it was a journey in the rust types to understand where this information is stored.

Uriopass commented:
Added some tests and changed the way the lint looks. Now prints something like
```Rust
error: This closure returns the unit type which also implements PartialOrd.
  --> $DIR/unintentional_unit_return.rs:20:30
   |
LL |     structs.is_sorted_by_key(|s| {
   |                              ^^^
   |
help: Probably caused by this leading semicolon.
  --> $DIR/unintentional_unit_return.rs:21:24
   |
LL |         double(s.field);
   |                        ^
```

flip1995 commented on clippy_lints/src/unintentional_unit_return.rs:
```diff
@@ -0,0 +1,176 @@
+                                UNINTENTIONAL_UNIT_RETURN,
+                                span,
+                                &format!(
+                                    "This closure returns \
+                                   the unit type which also implements {}.",
```
```suggestion
                                    "this closure returns \
                                   the unit type which also implements {}",
```

rust error/help/suggestion messages start lower case and don't have punctuation at the end. Please apply this everywhere.

flip1995 commented on clippy_lints/src/unintentional_unit_return.rs:
```diff
@@ -0,0 +1,176 @@
+fn check_arg<'tcx>(cx: &LateContext<'tcx>, arg: &'tcx Expr<'tcx>) -> Option<(Span, Option<Span>)> {
+    if_chain! {
+        if let ExprKind::Closure(_, _fn_decl, body_id, span, _) = arg.kind;
+        if let ty::Closure(_def_id, substs) = &cx.tables().node_type(arg.hir_id).kind;
+        let ret_ty = substs.as_closure().sig().output();
+        let ty = cx.tcx.erase_late_bound_regions(&ret_ty);
+        if ty.is_unit();
```
TIL, I suspected that it could work that way, but now I know for sure. Thanks! :)

flip1995 commented on tests/ui/unintentional_unit_return.rs:
```diff
@@ -0,0 +1,33 @@
+    });
+    structs.sort_by_key(|s| {
+        return double(s.field);
+    });
```
```suggestion
    });
    structs.sort_by_key(|s| {
        unit(s.field) // where unit is `fn unit(_: isize) {}`
    });
```

This test should Just Work™️, but I would add it anyway.

flip1995 commented:
Can you squash some of the commits, please? After that, this should be good to go 🚀

Uriopass commented:
Just, should this be in the nursery ?

flip1995 commented on clippy_lints/src/unintentional_unit_return.rs:
````diff
@@ -0,0 +1,176 @@
+    /// twins.sort_by_key(|x| { x.1; });
+    /// ```
+    pub UNINTENTIONAL_UNIT_RETURN,
+    nursery,
````
I think we can move this to `correctness` since it should be finished now. Or do you know of any problems you encountered with this lint, while implementing/testing it?

Uriopass replied:
Nope, I think it serves the exact purpose it was made for.

Uriopass commented:
Moved from nursery to correctness and rewritten to description to match current (updated) behavior

flip1995 commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 0f7cbf4ac901b2bc9d2472ad6f871f31c5814244 has been approved by `flip1995`

<!-- @bors r=flip1995 0f7cbf4ac901b2bc9d2472ad6f871f31c5814244 -->
<!-- homu: {"type":"Approved","sha":"0f7cbf4ac901b2bc9d2472ad6f871f31c5814244","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 0f7cbf4ac901b2bc9d2472ad6f871f31c5814244 with merge 88388ebcdfba2bdce800094c04a81a102cf4cee3...
<!-- homu: {"type":"BuildStarted","head_sha":"0f7cbf4ac901b2bc9d2472ad6f871f31c5814244","merge_sha":"88388ebcdfba2bdce800094c04a81a102cf4cee3"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/865622418)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/865622418","builder_name":"checks-action_test"} -->

flip1995 commented on clippy_lints/src/unintentional_unit_return.rs:
````diff
@@ -0,0 +1,177 @@
+    /// let mut twins = vec!((1,1), (2,2));
+    /// twins.sort_by_key(|x| { x.1; });
+    /// ```
+    pub UNINTENTIONAL_UNIT_RETURN,
````
I think the lint name doesn't completely represent what this lint is doing. IMO, it suggests that it may check for return values of type unit, that are unintentional. But what it rather does is check for unit returns where a `ORD` type is expected (like correctly described above). How about the following name:

```suggestion
    pub UNIT_RETURN_EXPECTING_ORD,
```

Uriopass commented:
Renamed and rustuped

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 126790999a128a880ca276c49afd2927a66ffbbe has been approved by `flip1995`

<!-- @bors r=flip1995 126790999a128a880ca276c49afd2927a66ffbbe -->
<!-- homu: {"type":"Approved","sha":"126790999a128a880ca276c49afd2927a66ffbbe","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 126790999a128a880ca276c49afd2927a66ffbbe with merge 97d7a8b319ea266da57a7964719a006582871be4...
<!-- homu: {"type":"BuildStarted","head_sha":"126790999a128a880ca276c49afd2927a66ffbbe","merge_sha":"97d7a8b319ea266da57a7964719a006582871be4"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/870140268), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/870137696), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/870216639)
Approved by: flip1995
Pushing 97d7a8b319ea266da57a7964719a006582871be4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/870216639","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/870140268","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/870137696"},"merge_sha":"97d7a8b319ea266da57a7964719a006582871be4"} -->

Uriopass commented:
Nice !
